### PR TITLE
Support for Conditions in VPA.

### PR DESCRIPTION
**Pull request by kgrygiel**

### CODE REVIEW

**Comment by mwielgus** [comment]
!!! 'gofmt -s' needs to be run on the following files: 
./vertical-pod-autoscaler/pkg/utils/vpa/api.go

File: vertical-pod-autoscaler/pkg/utils/vpa/api.go [hunk 1]
```diff
@@ -49,32 +50,34 @@ func patchVpa(vpaClient vpa_api.VerticalPodAutoscalerInterface, vpaName string,
...
-		Value: vpa_types.VerticalPodAutoscalerStatus{},
-	},
-	})
-}
+// UpdateVpaStatus updates the status field of the VPA API object.
+// It prevents race conditions by verifying that the lastUpdateTime of the
+// API object and its model representation are equal.
+func UpdateVpaStatus(vpaClient vpa_api.VerticalPodAutoscalerInterface, vpa *model.Vpa) (result *vpa_types.VerticalPodAutoscaler, err error) {
+	status := vpa_types.VerticalPodAutoscalerStatus{
+		LastUpdateTime: metav1.Now(),
```
Comment (tkulczynski) [inline vertical-pod-autoscaler/pkg/utils/vpa/api.go]: To be perfectly sure this solves races, you need to check that each patch strictly increases LastUpdateTime. While this should be true in sane worlds (but not e.g. if different nodes in the cluster have desynchronized clocks), I would add the if statement .

File: vertical-pod-autoscaler/pkg/utils/vpa/api.go [hunk 2]
```diff
@@ -49,32 +50,34 @@ func patchVpa(vpaClient vpa_api.VerticalPodAutoscalerInterface, vpaName string,
...
+		status.Recommendation = *vpa.Recommendation
+	}
 
-// UpdateVpaRecommendation updates VPA's status/recommendation object and status/lastUpdateTime.
-func UpdateVpaRecommendation(vpaClient vpa_api.VerticalPodAutoscalerInterface, vpaName string, recommendation vpa_types.RecommendedPodResources) (result *vpa_types.VerticalPodAutoscaler, err error) {
-	patches := []patchRecord{
-		{
-			Op:    "add",
+	patches := make([]patchRecord, 0)
+	if !vpa.LastUpdateTime.IsZero() {
```
Comment (tkulczynski) [inline vertical-pod-autoscaler/pkg/utils/vpa/api.go]: Why not test in that case? There's a race for setting the first status then, no?

Comment (kgrygiel) [reply vertical-pod-autoscaler/pkg/utils/vpa/api.go]: Unfortunately I cannot set a test on a missing field in the patch method.
On the other hand we can make the assumption that the Updater (or other agents) don't do any operations on a VPA that doesn't have the Status field. In particular it has no recommendation at this point, so the Updater can't do anything useful. WDYT?

Comment (kgrygiel) [reply vertical-pod-autoscaler/pkg/utils/vpa/api.go]: Done (discussed offline, if the lastUpdateTime is missing I set a test on the whole status field).

File: vertical-pod-autoscaler/pkg/utils/vpa/api.go (hunk 1 above)
Comment (kgrygiel) [reply vertical-pod-autoscaler/pkg/utils/vpa/api.go]: Done.

**Comment by tkulczynski** [comment]
LGTM.

@bskiba: could you approve?

**Pull request merged**